game-app/game-headed/src/main/java/games/strategy/triplea/ui/PlacePanel.java

```diff
@@ -145,15 +146,23 @@ private void updateStep() {
       }
     }
 
-    SwingUtilities.invokeLater(
-        () -> {
-          if (showUnitsToPlace) {
-            unitsToPlacePanel.setUnits(unitsToPlace);
+    if (showUnitsToPlace) {
+      // Small hack: copy the unit list before passing it to a new thread.
+      // This is to prevent ConcurrentModification. If the 'unitsToPlace' list is modified
+      // later in this thread, before "SwingUtilities.invokeLater" can execute and complete,
```

Comment:
I don't think this comment is quite right. If it's modified before invokeLater, that's fine.

I think the issue is if it's being modified while the swing thread is iterating over it.

I think this is actually the right fix, so I don't agree with the end of the comment "Ideally we would not modify...". I don't the an API should place a restriction on the caller to not modify something that's passed to it just because its implementation does something async to it. If it does something async, it seems appropriate to make a copy.

Reply:
Regarding time of modification:
- either there is a dirty flag that is set as soon as any modification is done (before we grab an iterator)
- or there is a dirty flag that is set after the iterator is grabbed

I think it is the first. I don't have time to research that right now. But the answer would be helpful.

As far as ownership & brittleness. Once the data is passed to a new thread - it seems like that new thread should then own all access & modifications to that data. What is the original thread doing modifying that data? That will create an inconsistent state between the other thread - it's a terribly brittle design. It would be less brittle if the data were immutable, and then any callbacks were owned by the new thread. 

I am guessing that at some point the entire flow was single threaded. That everything was done on the EDT, at some point that was corrected; but then just this last bit was put back on the EDT. Hence the threading issue was presented while still having this very odd flow of control.